[3.x] Add a Resource-Prefix Config Option to CLI

himani2411 opened the pull request:
### Description of changes

* Providing Resource Prefix config option so that user is able to customize the IAM resource's path prefix and role based prefix.
* Changing schema `ClusterIamSchema` and resource `ClusterIam` to enable user to give ResourcePrefix config option.
* Using IAM naming conventions to create a Regex which will validate the input.

### Tests
* Manual Tests

      1. Created a cluster with wrong Resource Prefix format to see the validation ERROR
      2. Created a cluster with wrong Resource Prefix format and suppressed validators to see the no validation ERROR
      3. Created a Cluster with correct Resource Prefix format (this wont be used for resource creation yet)

* Automated Tests

      1. Unit test cases to validate the Regex handles both positive and negative test scenarios.

### References
* [IAM Naming Conventions](https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_iam-quotas.html)

### Checklist
- [x] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

Commit:
Add Iam Resource Prefix as config option in schema.

Add Regex to handle the IAM naming conventions for path and Role Prefix.
Unit Test Cases for positive and negative inputs.

Commit:
Change Update Policy of Iam Resource Prefix

Keeping the Update Policy as Unsupported when adding Iam Resource Prefix

Commit:
Change IAM Resource Prefix as Schema Config option

Revert Regex applied on Iam Schema
Removed Unit tests

Commit:
Add resource prefix at resource level

Add regex to handle IAM naming conventions for Path and role prefix
Add a Validator and Unit Tests for different test cases
Add ResourcePrefix in YAML file for all validators tests

gmarciani commented on cli/tests/pcluster/schemas/test_schema_validators.py:
```diff
@@ -589,6 +589,7 @@
                     contain = True
         assert_that(contain).is_true()
     else:
+        print(section_dict)
```
We should remove the above `print`. Probably put there for debugging during development

gmarciani commented on cli/src/pcluster/validators/iam_validators.py:
```diff
@@ -57,6 +59,33 @@ def _validate(self, policy: str):
+    """
+    Iam Resource Prefix Validator.
+
+    Verify if the Resource Prefix belongs in the permission boundary if given.
```
I think this comment is wrong. The goal of this validator is to verify that the ResourcePrefix is compliant with the expected format

himani2411 replied:
I will change the Comment.

gmarciani commented on cli/src/pcluster/validators/iam_validators.py:
```diff
@@ -57,6 +59,33 @@ def _validate(self, policy: str):
+                or re.match(IamResourcePrefixValidator.IAM_ROLE_NAME_PREFIX, resource_prefix)
+            ):
+                self._add_failure(
+                    f"Resource Prefix {resource_prefix} provided does not fall under the accepted pattern",
```
Is the error message approved by PM?
I'd rephrase with something like "Unsupported format for ResourcePrefix {resource_prefix} " or "ResourcePrefix {resource_prefix} does not comply with expected format."

Furthermore, I'd provide some suggestions to our customers about the expected format.
For example, we could suggest them to have a look at IAM documentation

himani2411 replied:
No the message wasnt approved. I will take your suggestion to change it.

himani2411 replied:
The IAM documentation would be confusing right now. I was thinking of maybe later adding the link to our documentation where we mention the exact requirements.

gmarciani replied:
> I was thinking of maybe later adding the link to our documentation where we mention the exact requirements.

I agree. Let's invite the user to refer to our docs. 
In general, I think we should avoid putting specific links in error message because we would couple our code to something that could change

gmarciani replied:
I propose the message: "ResourcePrefix {resource_prefix} is not valid. Please, refer to our public docs for further details."

himani2411 replied:
Changed the Failure message as requested.

gmarciani commented on cli/src/pcluster/validators/iam_validators.py:
```diff
@@ -57,6 +59,33 @@ def _validate(self, policy: str):
+                    f"Resource Prefix {resource_prefix} provided does not fall under the accepted pattern",
+                    FailureLevel.ERROR,
+                )
+        except Exception as e:
```
What exception do you expect here?

himani2411 replied:
I have removed the Exception section it was to handle one of the test cases where I had kept the resource prefix as None, however that condition is handled even before we call the Validation.

gmarciani commented on cli/src/pcluster/validators/iam_validators.py:
```diff
@@ -57,6 +59,33 @@ def _validate(self, policy: str):
+                )
+        except Exception as e:
+            self._add_failure(
+                f"Resource Prefix {resource_prefix} provided does not fall under the accepted pattern",
```
This error message is specifically meant for ResourcePrefix being in an invalid format.
If we have an exception here, it does not necessarily mean that. Right?

himani2411 replied:
I have removed the Exception section it was to handle one of the test cases where I had kept the resource prefix as None, however that condition is handled even before we call the Validation.

gmarciani reviewed (commented):
There are some failing checks. May you please fix? Thanks

Commit:
Change w.r.t PR for ResourcePrefix option

Remove Exception Handler as it was handling a Unit Test case which is handled before calling Validation.
Removed Print statement which was for debugging and understanding the code
PR Link:
https://github.com/aws/aws-parallelcluster/pull/4524

Commit:
Change Error message for ResourcePrefix option

Change the Failure message for Resource Prefix redirecting  to our public Documentation.

PR Link:
https://github.com/aws/aws-parallelcluster/pull/4524

gmarciani commented:
In this PR we are introducing a new cluster config parameter.
This type of change should be reported into our CHANGELOG.
May you please add it?
Thanks

gmarciani commented on cli/src/pcluster/validators/iam_validators.py:
```diff
@@ -57,6 +59,28 @@ def _validate(self, policy: str):
+        ):
+            self._add_failure(
+                f"Unsupported format for ResourcePrefix {resource_prefix} "
+                f"Please, refer to our public docs for further details.",
```
1. I'd say "our official documentation" instead of "public docs" because the latter sounds too "internal".
2. Missing full stop at the end of the first sentence.

himani2411 replied:
Sure

gmarciani commented on cli/tests/pcluster/validators/test_all_validators/test_scheduler_plugin_all_validators_are_called/scheduler_plugin_1.yaml:
```diff
@@ -235,6 +235,7 @@ SharedStorage:
       StorageType: HDD  # HDD | SSD
 Iam:
   PermissionsBoundary: arn:aws:iam::aws:policy/boundary
+  ResourcePrefix: /boundary/role-prefix
```
**General best practice**
I suggest to replace `/boundary/role-prefix` with something like `/path/role-prefix` or `/whatever-path/whatever-role-prefix` because we should not suggest the idea that the prefix used in `ResourcePrefix` should match the prefix used by the `PermissionBoundary`.

Let's apply this best practices in every other test parameter in this PR.

himani2411 replied:
I will change it to /path-prefix/role-prefix

gmarciani replied:
Not clear to me: when we put `ResourcePrefix` is `/blablabla/tiptoe`, then `blablabla` will be the path or the path prefix?

himani2411 replied:
Will rename to /path-prefix/name-prefix

gmarciani commented on cli/tests/pcluster/validators/test_iam_validators.py:
```diff
@@ -113,3 +115,125 @@ def test_additional_iam_policy_validator(mocker, policy_arn, expected_get_policy
+
+
+@pytest.mark.parametrize(
+    "resource_prefix, expected_message,expected_failure_level",
```
We always expect the failure level to be `FailureLevel.ERROR` if a failure occurs. 
so, we can avoid the parameter `expected_failure_level` and assert `FailureLevel.ERROR` in the logic directly.

himani2411 replied:
Changing this string would lead me to make changes on the Function level as well, which would give an unresolved error as it wont compare with the parameter we are giving but the defined Enum we have.

himani2411 replied:
After a call, we have resolved to keep the test function as it is.

gmarciani commented on cli/tests/pcluster/validators/test_iam_validators.py:
```diff
@@ -113,3 +115,125 @@ def test_additional_iam_policy_validator(mocker, policy_arn, expected_get_policy
+    [
+        (
+            r"\pathprefix\\",
+            "Unsupported format for ResourcePrefix",
```
If possible, it's always better to test the entire error message that is returned, not just a part of it.

himani2411 replied:
This case was giving an exception while comparing due to the the escape character given and which is part of the expected message.

himani2411 replied:
Will add a comment for TO DO

gmarciani commented on cli/tests/pcluster/validators/test_iam_validators.py:
```diff
@@ -113,3 +115,125 @@ def test_additional_iam_policy_validator(mocker, policy_arn, expected_get_policy
+        ("/pathprefix/", None, None),
+        ("role-prefix", None, None),
+        ("@roleprefix", None, None),
+        ("role-prefix", None, None),
```
duplicated case

himani2411 replied:
Remved Duplication

gmarciani commented on cli/tests/pcluster/validators/test_iam_validators.py:
```diff
@@ -113,3 +115,125 @@ def test_additional_iam_policy_validator(mocker, policy_arn, expected_get_policy
+        ("/pathprefix/parallelcluster/", None, None),
+        ("/parallelcluster/pathprefix/", None, None),
+        ("/pathprefix/", None, None),
+        ("role-prefix", None, None),
```
We can reduce the number of cases to the significative ones.
It would make sense to have one case including all the special characters that are accepted.
For example: `role-prefix@-+=._,`

himani2411 replied:
Sure

gmarciani commented on cli/tests/pcluster/validators/test_iam_validators.py:
```diff
@@ -113,3 +115,125 @@ def test_additional_iam_policy_validator(mocker, policy_arn, expected_get_policy
+        ("/path-prefix/role-prefix", None, None),
+        ("/path@prefix/+roleprefix/", None, None),
+        (
+            "/path+prefix/=roleprefix",
```
Minor: why the below cases not in single lines?

himani2411 replied:
Will change that

gmarciani commented on cli/tests/pcluster/validators/test_iam_validators.py:
```diff
@@ -113,3 +115,125 @@ def test_additional_iam_policy_validator(mocker, policy_arn, expected_get_policy
+            FailureLevel.ERROR,
+        ),
+        (
+            "&&&roleprefix",
```
It's not just role name prefix but also policy name prefix, right?

himani2411 replied:
No, the user can below inputs
-> Both Path prefix and role name prefix
-> Just Path prefix
-> Just role name prefix

himani2411 replied:
Inputs

Both: ( Start with / but Shouldnt end with / )
/pathprefix/roleprefix 
Just Path prefix: ( Start and end with /)
/pathprefix/
Just Rolename: (No /)
roleprefix

himani2411 replied:
Will rename this is to name prefix

gmarciani commented on cli/tests/pcluster/validators/test_iam_validators.py:
```diff
@@ -113,3 +115,125 @@ def test_additional_iam_policy_validator(mocker, policy_arn, expected_get_policy
+            FailureLevel.ERROR,
+        ),
+        (
+            ",./pathprefix/",
+            "Unsupported format for ResourcePrefix ,./pathprefix/",
+            FailureLevel.ERROR,
+        ),
+        (
+            ";)*/anything/pathprefix/",
+            "Unsupported format for ResourcePrefix",
+            FailureLevel.ERROR,
+        ),
+        (
+            "#$/pathprefix/anything/",
+            "Unsupported format for ResourcePrefix ",
+            FailureLevel.ERROR,
+        ),
```
We can collapse these cases into a single one

himani2411 replied:
Sure

gmarciani commented on cli/src/pcluster/validators/iam_validators.py:
```diff
@@ -57,6 +59,28 @@ def _validate(self, policy: str):
+    """
+
+    IAM_PATH_NAME_REGEX = r"^((/[a-zA-Z0-9_.,+@=-]+)+)/"
+    IAM_ROLE_NAME_PREFIX = r"^[a-zA-Z0-9_.,+@=-]+$"
```
It's not just a role name prefix, but also policy name prefix, right?

himani2411 replied:
https://github.com/aws/aws-parallelcluster/pull/4524#discussion_r1019246026

himani2411 replied:
Will rename IAM_ROLE_NAME_PREFIX to IAM_NAME_PREFIX_REGEX

Commit:
Update ChangeLog to add Resource Prefix enhancement

Update in CHANGELOG to mention new enhancement for Resource Prefix Config option in 3.4
Change parameter names to match the config option names
Fix Test Cases to remove duplicate cases and combine multiple cases

PR Link:
https://github.com/aws/aws-parallelcluster/pull/4524

gmarciani commented on CHANGELOG.md:
```diff
@@ -1,5 +1,10 @@
+-----
+
+**ENHANCEMENTS**
+- Add new configuration parameter `ResourcePrefix` in `Iam` to specify a Path prefix and Name prefix for IAM resources.
```
According to our best practices the line should be something like:

"Add new configuration parameter in `Iam/ResourcePrefix` to specify a prefix for path and name of IAM resources created by ParallelCluster."

himani2411 replied:
Made changes

Commit:
Update ChangeLog to correct Resource Prefix enhancement statement

Update in CHANGELOG to mention new enhancement for Resource Prefix Config option in 3.4

PR Link:
https://github.com/aws/aws-parallelcluster/pull/4524

dreambeyondorange committed:
Merge branch 'develop' into wip/iam_resource_prefix

dreambeyondorange committed:
Merge branch 'develop' into wip/iam_resource_prefix

himani2411 committed:
Merge branch 'develop' into wip/iam_resource_prefix
